common/hashing_test.cpp

```diff
@@ -0,0 +1,589 @@
+  }
+
+  for (int i : {-1, -2, -3, -13}) {
+    SCOPED_TRACE(llvm::formatv("Hashing: {0}", i).str());
+
+    // Negative numbers can't be cheaply made to hash consistently regardless of
+    // size -- doing so would force sign extensions that are often expensive.
+    // Instead, we can check that the exact 2s compliment form at the bit-width
+    // of the signed integer is used.
+    EXPECT_THAT(HashValue(static_cast<int8_t>(i)),
+                Eq(HashValue(static_cast<uint8_t>(static_cast<int8_t>(i)))));
+    EXPECT_THAT(HashValue(static_cast<int16_t>(i)),
+                Eq(HashValue(static_cast<uint16_t>(static_cast<int16_t>(i)))));
+    EXPECT_THAT(HashValue(static_cast<int32_t>(i)),
+                Eq(HashValue(static_cast<uint32_t>(static_cast<int32_t>(i)))));
+    EXPECT_THAT(HashValue(static_cast<int64_t>(i)),
+                Eq(HashValue(static_cast<uint64_t>(static_cast<int64_t>(i)))));
+  }
+}
+
+TEST(HashingTest, Pointers) {
+  int object1 = 42;
+  std::string object2 =
+      "Hello World! This is a long-ish string so it ends up on the heap!";
+
+  HashCode hash_null = HashValue(nullptr);
+  // Hashes should be stable.
+  EXPECT_THAT(HashValue(nullptr), Eq(hash_null));
+
+  // Hash other kinds of pointers without trivial collisions.
+  HashCode hash1 = HashValue(&object1);
+  HashCode hash2 = HashValue(&object2);
+  HashCode hash3 = HashValue(object2.data());
+  EXPECT_THAT(hash1, Ne(hash_null));
+  EXPECT_THAT(hash2, Ne(hash_null));
+  EXPECT_THAT(hash3, Ne(hash_null));
+  EXPECT_THAT(hash1, Ne(hash2));
+  EXPECT_THAT(hash1, Ne(hash3));
+  EXPECT_THAT(hash2, Ne(hash3));
+
+  // Hash values reflect the address and not the type.
+  EXPECT_THAT(HashValue(static_cast<void*>(nullptr)), Eq(hash_null));
+  EXPECT_THAT(HashValue(static_cast<int*>(nullptr)), Eq(hash_null));
+  EXPECT_THAT(HashValue(static_cast<std::string*>(nullptr)), Eq(hash_null));
+  EXPECT_THAT(HashValue(reinterpret_cast<void*>(&object1)), Eq(hash1));
+  EXPECT_THAT(HashValue(reinterpret_cast<int*>(&object2)), Eq(hash2));
+  EXPECT_THAT(HashValue(reinterpret_cast<std::string*>(object2.data())),
+              Eq(hash3));
+}
+
+TEST(HashingTest, PairsAndTuples) {
+  // Note that we can't compare hash codes across arity, or in general, compare
+  // hash codes for different types as the type isn't part of the hash.
+  HashCode hash_2zero = HashValue(std::pair(0, 0));
+  EXPECT_THAT(HashValue(std::pair(0, 1)), Ne(hash_2zero));
+  EXPECT_THAT(HashValue(std::pair(1, 0)), Ne(hash_2zero));
+
+  HashCode hash_3zero = HashValue(std::tuple(0, 0, 0));
+  EXPECT_THAT(HashValue(std::tuple(0, 0, 1)), Ne(hash_3zero));
+  EXPECT_THAT(HashValue(std::tuple(0, 1, 0)), Ne(hash_3zero));
+  EXPECT_THAT(HashValue(std::tuple(1, 0, 0)), Ne(hash_3zero));
+
+  // Hashing a 2-tuple and a pair should produce identical results though as we
+  // want to be able to have things like variadic tuple construction be
+  // indistinct from pairs.
```

Comment:
Could you perhaps rephrase to avoid using "indistinct"? Maybe:
```suggestion
  // Hashing a 2-tuple and a pair should produce identical results, so pairs
  // are compatible with code using things like variadic tuple construction.
```